README.md
coinkey
=======

[![Version](http://img.shields.io/npm/v/coinkey.svg)](https://www.npmjs.org/package/coinkey)
[![build status](https://secure.travis-ci.org/cryptocoinjs/coinkey.png)](http://travis-ci.org/cryptocoinjs/coinkey)
[![Coverage Status](https://img.shields.io/coveralls/cryptocoinjs/coinkey.svg)](https://coveralls.io/r/cryptocoinjs/coinkey)

JavaScript component for private keys, public keys, and addresses for crypto currencies such as Bitcoin, Litecoin, and Dogecoin. Works
in both Node.js and the browser.



Package Info
------------
- github: [https://github.com/cryptocoinjs/coinkey](https://github.com/cryptocoinjs/coinkey)
- tests: [https://github.com/cryptocoinjs/coinkey/tree/master/test](https://github.com/cryptocoinjs/coinkey/tree/master/test)
- issues: [https://github.com/cryptocoinjs/coinkey/issues](https://github.com/cryptocoinjs/coinkey/issues)
- license: **MIT**
- versioning: [http://semver-ftw.org](http://semver-ftw.org)


Installation
------------

    npm i --save coinkey


Usage
-----

### Common Use Cases

### Generate a Bunch of Bitcoin Keys/Addresses

```js
var CoinKey = require('coinkey')

var bitcoinKeys = []
for (var i = 0; i < 10; ++i) {
  // bitcoin supported by default
  bitcoinKeys.push(CoinKey.createRandom())
}
```


#### Generate a Bunch of Namecoin Keys/Addresses

```js
var CoinKey = require('coinkey')
// npm install --save coininfo
var ci = require('coininfo')

var namecoins = []
for (var i = 0; i < 10; ++i) {
  namecoins.push(CoinKey.createRandom(ci('NMC')))
}
```


#### Parse a Wallet Import Key and Determine Crypto Currency

```js
var CoinKey = require('coinkey')
var ci = require('coininfo')

var ck = CoinKey.fromWif('QVD3x1RPiWPvyxbTsfxVwaYLyeBZrQvjhZ2aZJUsbuRgsEAGpNQ2')

console.log(ck.privateKey.toString('hex')) // => c4bbcb1fbec99d65bf59d85c8cb62ee2db963f0fe106f483d9afa73bd4e39a8a
console.log(ck.publicAddress) // => DGG6AicS4Qg8Y3UFtcuwJqbuRZ3Q7WtYXv
console.log(ck.compressed) // => true
console.log(ck.versions.public === ci('DOGE').versions.public) // => true
```


#### Change to Testnet Later

```js
var CoinKey = require('coinkey')
var ci = require('coininfo')

var ck = new CoinKey(new Buffer('1184cd2cdd640ca42cfc3a091c51d549b2f016d454b2774019c2b2d2e08529fd', 'hex'))
console.log(ck.publicAddress) // => 16UjcYNBG9GTK4uq2f7yYEbuifqCzoLMGS

//change to Testnet
ck.versions = ci('BTC-TEST')

console.log(ck.publicAddress) // => mkzgubTA5Ahi6BPSkE6MN9pEafRutznkMe
```


### API

#### CoinKey(privateKey, [versions])

Constructor function.

- **privateKey**: The private key bytes. Must be 32 bytes in length. Should be an `Array`, `Uint8Array`, or a `Buffer`.
- **versions**: An object that specifies the public and private key versions for addresses and wifs. Defaults to Bitcoin `mainnet`.

Keys are default set to `compressed` is `true`.

```js
var CoinKey = require('coinkey')
//npm install --save secure-random@1.x
var secureRandom = require('secure-random')

var bytes = secureRandom.randomBuffer(32)
var key1 = new CoinKey(bytes)
console.log(key1.compressed) // => true
```


### Properties


#### compressed

Inherited from [ECKey][eckey]. [eckey.compressed](http://cryptocoinjs.com/modules/currency/eckey/#compressed)


#### privateKey

Inherited from [ECKey][eckey]. [eckey.privateKey](http://cryptocoinjs.com/modules/currency/eckey/#privatekey)


#### privateExportKey

Inherited from [ECKey][eckey]. [eckey.privateExportKey](http://cryptocoinjs.com/modules/currency/eckey/#privateexportkey)


#### privateWif

Get the private WIF (Wallet Import Format).

```js
var CoinKey = require('coinkey')

var privateKeyHex = "1184cd2cdd640ca42cfc3a091c51d549b2f016d454b2774019c2b2d2e08529fd"

//Bitcoin WIF
var key = new CoinKey(new Buffer(privateKeyHex, 'hex'))
key.compressed = false
console.log(key.privateWif) // => 5Hx15HFGyep2CfPxsJKe2fXJsCVn5DEiyoeGGF6JZjGbTRnqfiD

//Litecoin WIF
var key = new CoinKey(new Buffer(privateKeyHex, 'hex'), {private: 0xB0, public: 0x30})
key.compressed = false
console.log(key.privateWif) // => 6uFjYQnot5Gtg3HpP87bp4JUpg4FH1gkkV3RyS7LHBbD9Hpt1na
```


#### publicKey

Inherited from [ECKey][eckey]. [eckey.publicKey](http://cryptocoinjs.com/modules/currency/eckey/#publickey)


#### publicAddress

Get the public address.

```js
var CoinKey = require('coinkey')

var privateKeyHex = "1184cd2cdd640ca42cfc3a091c51d549b2f016d454b2774019c2b2d2e08529fd"

// Bitcoin Address
var key = new CoinKey(new Buffer(privatKeyHex, 'hex'))
console.log(key.publicAddress) // => 16UjcYNBG9GTK4uq2f7yYEbuifqCzoLMGS

// Litecoin Address
var key = new CoinKey(new Buffer(privateKeyHex, 'hex'), {private: 0xB0, public: 0x30})
console.log(key.publicAddress) // => 'LZyGd5dCPVkVUjA5QbpuUfMNgcmNDLjswH'
```


#### publicHash

Alias: `pubKeyHash`

Inherited from [ECKey][eckey]. [eckey.publicHash](http://cryptocoinjs.com/modules/currency/eckey/#publicHash)


#### publicPoint

Inherited from [ECKey][eckey]. [eckey.publicPoint](http://cryptocoinjs.com/modules/currency/eckey/#publicPoint)


#### toString()

Returns the string representation.

```js
var CoinKey = require('coinkey')

var privateKeyHex = "1184cd2cdd640ca42cfc3a091c51d549b2f016d454b2774019c2b2d2e08529fd"

//Litecoin Address
var key = new CoinKey(new Buffer(privateKeyHex, 'hex'), {private: 0xB0, public: 0x30})
console.log(key.toString())
// => T3e2me1BvRs95K7E8eQ8eha9oRPL1g2U6vmjE5px6RjzbUTvKZsf: LZyGd5dCPVkVUjA5QbpuUfMNgcmNDLjswH
```


### Methods

#### fromWif(wif, [versions])

Class method to create a `CoinKey` from a wif.

```js
var ck = CoinKey.fromWif('KwomKti1X3tYJUUMb1TGSM2mrZk1wb1aHisUNHCQXTZq5auC2qc3');
console.log(ck.compressed) // => true
console.log(ck.privateKey.toString('hex')) // => 1184cd2cdd640ca42cfc3a091c51d549b2f016d454b2774019c2b2d2e08529fd
console.log(ck.publicAddress) // => 1FkKMsKNJqWSDvTvETqcCeHcUQQ64kSC6s
```


Browser Support
---------------

Clone the repo:

    git clone https://github.com/cryptocoinjs/coinkey

Install Browserify

    npm install -g browserify

Nav to repo:

    cd coinkey

Install dependencies:

    npm install

Run browserify:

    browserify --standalone coinkey lib/coinkey.js > lib/coinkey.bundle.js

You can now drop `coinkey.bundle.js` in a `<script>` tag.


[eckey]: https://github.com/cryptocoinjs/eckey
[coinstring]: https://github.com/cryptocoinjs/coinstring


Hack on CoinKey
---------------

[![js-standard-style](https://cdn.rawgit.com/feross/standard/master/badge.svg)](https://github.com/feross/standard)


CHANGELOG.md
2.0.0 / 2016-04-27
------------------
- upgraded from [`ecurve`](https://www.npmjs.com/package/ecurve) to [`secp256k1`](https://www.npmjs.com/package/secp256k1) (via `eckey` upgrade).

1.4.0 / 2015-04-20
------------------
- JavaScript Standard Style
- support [coininfo](https://github.com/cryptocoinjs/coininfo) objects without having to specify `versions`
- added method `addressToHash()`
- brought back `README` from http://cryptocoinjs.com

1.3.0 / 2014-09-29
------------------
* upgraded `"eckey": "^0.7.0"` to `"eckey": "^0.8.0"` (which upgraded ecurve from 0.8.0 to 1.0.0; removed superflous params)

1.2.0 / 2014-06-30
------------------
* upgraded `"coinstring": "^1.0.1"` to `"coinstring": "^2.0.0"`
* added `crypto-browserify` for Browserify build and `make browser-test`

1.1.0 / 2014-06-11
------------------
* upgraded `"eckey": "^0.6.0"` to `"eckey": "^0.7.0"`

1.0.0 / 2014-06-07
-------------------------
* added `mochify` as a dev dep
* removed semicolons per http://cryptocoinjs.com/about/contributing/#semicolons
* removed `terst` and moved to `assert`
* upgraded `"eckey": "~0.4.0"` to `"eckey": "^0.6.0"`
* deleted fields `pubKeyHash` and alias `publicHash` because they're now present in inherited `ECKey`
* upgraded `"coinstring": "~0.2.0"` to `"coinstring": "^1.0.1"`
* upgraded `"secure-random": "~0.2.0"` to `"secure-random": "^1.0.0"`
* removed `CoinKey.compressByDefault`, default is now `true`
* removed ability for `CoinKey` constructor to generate random key when `privateKey` aren't passed
* removed `binstring` dev dep
* added Travis CI
* added Coveralls
* added new class method `createRandom()`
* changed constructor signature from `CoinKey([privateKey], [compresssed], [versions])` to `CoinKey(privateKey, [versions])`

0.1.0 / 2014-03-10
------------------
* added method `fromWif()`
* upgraded to `coinstring@0.2.x` => `decode()` method changed
* fixed bug when bytes wasn't passed, it should generate a private key

0.0.1 / 2014-03-09
------------------
* initial release


test/coinkey.test.js
var assert = require('assert')
var coininfo = require('coininfo')
var secureRandom = require('secure-random')
var CoinKey = require('../')
var fixtures = require('./fixtures/coinkey')

/* global describe, it */

describe('CoinKey', function () {
  describe('constructor', function () {
    describe('> when private key passed', function () {
      it('should return an instance of CoinKey with fields set', function () {
        var privateKey = secureRandom.randomBuffer(32)
        var ck = new CoinKey(privateKey)
        assert(ck.compressed)
        assert.equal(ck.privateKey.toString('hex'), privateKey.toString('hex'))
      })
    })

    describe('> when private key and versions', function () {
      it('should return an instance of CoinKey with versions', function () {
        var dogecoin = fixtures.valid.filter(function (f) { if (f.description.match(/dogecoin/)) return f})[0]
        var ck = new CoinKey(new Buffer(dogecoin.privateKey, 'hex'), dogecoin.versions)
        assert(ck.compressed)
      })
    })

    describe('> when no private key', function () {
      it('should throw an error', function () {
        /* eslint-disable no-new */
        assert.throws(function () { new CoinKey() }, /must be arrayish/)
        assert.throws(function () { new CoinKey({public: 0, private: 0x80}) }, /must be arrayish/)
        /* eslint-enable no-new */
      })
    })

    describe('> when a coininfo object is passed for versions', function () {
      it('should return the proper address / WIF', function () {
        var dogecoin = fixtures.valid.filter(function (f) { if (f.description.match(/dogecoin/)) return f})[0]
        var ck = new CoinKey(new Buffer(dogecoin.privateKey, 'hex'), coininfo('DOGE'))
        assert.equal(ck.privateWif, dogecoin.privateWifCompressed)
        assert.equal(ck.publicAddress, dogecoin.publicAddressCompressed)
      })
    })
  })

  describe('- privateWif', function () {
    fixtures.valid.forEach(function (f) {
      it('should return the proper wif for ' + f.description, function () {
        var ck = new CoinKey(new Buffer(f.privateKey, 'hex'), f.versions)
        ck.compressed = false
        assert.equal(ck.privateWif, f.privateWif)
        var ckCompressed = new CoinKey(new Buffer(f.privateKey, 'hex'), f.versions)
        assert.equal(ckCompressed.privateWif, f.privateWifCompressed)
      })
    })
  })

  describe('- publicAddress', function () {
    fixtures.valid.forEach(function (f) {
      it('should return the proper public address for ' + f.description, function () {
        var ck = new CoinKey(new Buffer(f.privateKey, 'hex'), f.versions)
        ck.compressed = false
        assert.equal(ck.publicAddress, f.publicAddress)
        var ckCompressed = new CoinKey(new Buffer(f.privateKey, 'hex'), f.versions)
        assert.equal(ckCompressed.publicAddress, f.publicAddressCompressed)
      })
    })
  })

  describe('- toString()', function () {
    fixtures.valid.forEach(function (f) {
      it('should return the string ' + f.description, function () {
        var ck = new CoinKey(new Buffer(f.privateKey, 'hex'), f.versions)
        ck.compressed = false
        assert.equal(ck.toString(), f.privateWif + ': ' + f.publicAddress)
        var ckCompressed = new CoinKey(new Buffer(f.privateKey, 'hex'), f.versions)
        assert.equal(ckCompressed.toString(), f.privateWifCompressed + ': ' + f.publicAddressCompressed)
      })
    })
  })

  describe('+ fromWif()', function () {
    fixtures.valid.forEach(function (f) {
      it('should return a new CoinKey ' + f.description, function () {
        var ck = CoinKey.fromWif(f.privateWif)
        assert.equal(ck.compressed, false)
        assert.equal(ck.versions.private, f.versions.private)
        assert.equal(ck.versions.public, f.versions.public)
        assert.equal(ck.privateKey.toString('hex'), f.privateKey)
        assert.equal(ck.publicAddress, f.publicAddress)

        var ckCompressed = CoinKey.fromWif(f.privateWifCompressed)
        assert.equal(ckCompressed.compressed, true)
        assert.equal(ckCompressed.versions.private, f.versions.private)
        assert.equal(ckCompressed.versions.public, f.versions.public)
        assert.equal(ckCompressed.privateKey.toString('hex'), f.privateKey)
        assert.equal(ckCompressed.publicAddress, f.publicAddressCompressed)
      })
    })
  })

  describe('- versions', function () {
    describe('> when object changes', function () {
      it('should change the wif and public address', function () {
        var bitcoin = fixtures.valid.filter(function (f) { if (f.description.match(/bitcoin/)) return f })[0]
        var dogecoin = fixtures.valid.filter(function (f) { if (f.description.match(/bitcoin/)) return f })[0]

        var ck = new CoinKey(new Buffer(bitcoin.privateKey, 'hex'))
        assert.equal(ck.privateWif, bitcoin.privateWifCompressed)
        assert.equal(ck.publicAddress, bitcoin.publicAddressCompressed)

        ck.versions = dogecoin.versions // change to DOGECOIN

        assert.equal(ck.privateWif, dogecoin.privateWifCompressed)
        assert.equal(ck.publicAddress, dogecoin.publicAddressCompressed)

      })
    })

    describe('> when field changes', function () {
      it('should change the wif and public address', function () {
        var bitcoin = fixtures.valid.filter(function (f) { if (f.description.match(/bitcoin/)) return f })[0]
        var dogecoin = fixtures.valid.filter(function (f) { if (f.description.match(/bitcoin/)) return f })[0]

        var ck = new CoinKey(new Buffer(bitcoin.privateKey, 'hex'))
        assert.equal(ck.privateWif, bitcoin.privateWifCompressed)
        assert.equal(ck.publicAddress, bitcoin.publicAddressCompressed)

        // change to DOGECOIN
        ck.versions.private = dogecoin.versions.private
        ck.versions.public = dogecoin.versions.public

        assert.equal(ck.privateWif, dogecoin.privateWifCompressed)
        assert.equal(ck.publicAddress, dogecoin.publicAddressCompressed)
      })
    })
  })

  describe('+ createRandom()', function () {
    describe('> when no versions', function () {
      it('should create a random Bitcoin CoinKey', function () {
        var ck = CoinKey.createRandom()
        assert(ck.compressed)
        assert(ck.privateKey)
      })
    })

    describe('> when versions', function () {
      it('should create a random CoinKey with versions specified', function () {
        var dogecoin = fixtures.valid.filter(function (f) { if (f.description.match(/bitcoin/)) return f })[0]
        var ck = CoinKey.createRandom(dogecoin.versions)
        assert(ck.compressed)
        assert(ck.privateKey)
        assert.equal(ck.versions.private, dogecoin.versions.private)
        assert.equal(ck.versions.public, dogecoin.versions.public)
      })
    })
  })

  describe('+ addressToHash()', function () {
    it('should convert an address to the hash', function () {
      assert.equal(CoinKey.addressToHash('1FkKMsKNJqWSDvTvETqcCeHcUQQ64kSC6s').toString('hex'), 'a1c2f92a9dacbd2991c3897724a93f338e44bdc1')
      assert.equal(CoinKey.addressToHash('NBKgZWpMEDbzkTiRWHABRASXCdo8zWM8qC').toString('hex'), 'a1c2f92a9dacbd2991c3897724a93f338e44bdc1')
      assert.equal(CoinKey.addressToHash('3NukJ6fYZJ5Kk8bPjycAnruZkE5Q7UW7i8').toString('hex'), 'e8c300c87986efa84c37c0519929019ef86eb5b4')
    })
  })
})


test/coinkey-util.test.js
var assert = require('assert')
var util = require('../lib/util')

/* global describe, it */

describe('util', function () {
  describe('normalizeVersions()', function () {
    it('should normalize the versions object (i.e. be compatible with coininfo)', function () {
      assert.deepEqual(util.normalizeVersions({ public: 1, private: 2 }), { public: 1, private: 2 })
      assert.deepEqual(util.normalizeVersions({ versions: { public: 1, private: 2 } }), { public: 1, private: 2 })
      assert.deepEqual(util.normalizeVersions({ version: { public: 1, private: 2 } }), { public: 1, private: 2 })

      assert.deepEqual(util.normalizeVersions({}), null)
      assert.deepEqual(util.normalizeVersions(null), null)
    })
  })
})


lib/util.js
// poor man's clone
function clone (obj) {
  return JSON.parse(JSON.stringify(obj))
}

function isArrayish (maybeArray) {
  return Array.isArray(maybeArray)
    || (maybeArray instanceof Uint8Array)
    || Buffer.isBuffer(maybeArray)
}

// versions === { private, public } ? if no, check '.version(s)'
function normalizeVersions (versions) {
  if (!versions) return null
  if (typeof versions !== 'object') return null
  versions = clone(versions)
  if (versions.version) versions.versions = versions.version
  // check for actual versions object
  if (versions && 'private' in versions) return versions
  // if it exists, maybe in .versions?
  else versions = versions.versions
  if (versions && 'private' in versions) return versions
  else return null
}

module.exports = {
  clone: clone,
  isArrayish: isArrayish,
  normalizeVersions: normalizeVersions
}


lib/coinkey.js
var assert = require('assert')
var cs = require('coinstring')
var ECKey = require('eckey')
var inherits = require('inherits')
var secureRandom = require('secure-random')
var util = require('./util')

// Bitcoin
var DEFAULT_VERSIONS = {
  public: 0x0,
  private: 0x80
}

function CoinKey (privateKey, versions) {
  if (!(this instanceof CoinKey)) return new CoinKey(privateKey, versions)
  assert(util.isArrayish(privateKey), 'privateKey must be arrayish')
  this._versions = util.normalizeVersions(versions) || util.clone(DEFAULT_VERSIONS)
  // true => default compressed
  ECKey.call(this, privateKey, true)
}
inherits(CoinKey, ECKey)

// Instance props

Object.defineProperty(CoinKey.prototype, 'versions', {
  enumerable: true, configurable: true,
  get: function () {
    return this._versions
  },
  set: function (versions) {
    this._versions = versions
  }
})

Object.defineProperty(CoinKey.prototype, 'privateWif', {
  get: function () {
    return cs.encode(this.privateExportKey, this.versions.private)
  }
})

Object.defineProperty(CoinKey.prototype, 'publicAddress', {
  get: function () {
    return cs.encode(this.pubKeyHash, this.versions.public)
  }
})

CoinKey.prototype.toString = function () {
  return this.privateWif + ': ' + this.publicAddress
}

// Class methods

CoinKey.fromWif = function (wif, versions) {
  versions = util.normalizeVersions(versions)
  var res = cs.decode(wif)
  var version = res.slice(0, 1)
  var privateKey = res.slice(1) // get rid of version byte
  var compressed = (privateKey.length === 33)
  if (compressed) privateKey = privateKey.slice(0, 32) // slice off compression byte

  var v = versions || {}
  v.private = v.private || version.readUInt8(0)
  v.public = v.public || v.private - 0x80

  var ck = new CoinKey(privateKey, v)
  ck.compressed = compressed
  return ck
}

CoinKey.createRandom = function (versions) {
  var privateKey = secureRandom.randomBuffer(32)
  return new CoinKey(privateKey, versions)
}

CoinKey.addressToHash = function (address) {
  return cs.decode(address).slice(1)
}

module.exports = CoinKey
